File: processLine.py
import sys
import helper

lineNumber = int(sys.argv[1])
parts = int(sys.argv[2])
gradleFile = sys.argv[3]
print('lineNumber:', lineNumber)
print('     parts:', parts)
print('gradleFile:', gradleFile)
print()

lines = open(gradleFile, "r").readlines()
line = lines[lineNumber]
print('line', line)

oldVersion = helper.matchVersion(line, True)
if not oldVersion:
    helper.fail('no old version')
tempVersion = helper.getTempVersion(oldVersion, parts)
print('tempVersion', tempVersion)
if not tempVersion:
    helper.fail('no temp version')

oldDependencies = helper.command("""
    dependenciesFile=./bot/dependancies
    if grep -q $(git rev-parse HEAD) $dependenciesFile; then
      echo 'dependancies hash hit'
    else
      echo 'dependancies hash miss'
      ./gradlew vapi-service:dependencies --no-daemon > $dependenciesFile
      echo "$(git rev-parse HEAD)" >> $dependenciesFile
    fi
    cat $dependenciesFile
""").split('\n')

newVersion = None

if 'testImplementation' in line or 'implementation' in line or 'compile' in line:
    f = "./bot/parts{0}".format(parts)
    newDependencies = open(f, "r").readlines()
    newVersion = helper.lookForFirstOccurance(newDependencies, line)
else:
    lines[lineNumber] = line.replace(oldVersion, tempVersion)
    open(gradleFile, "w").writelines(lines)
    newDependencies = helper.command("./gradlew vapi-service:dependencies --no-daemon").split('\n')
    newVersion = helper.lookForFirstDiffVersion(oldDependencies, newDependencies)

print('newVersion', newVersion)
if (not helper.isNewVersionActuallyNewer(newVersion, oldVersion)):
    helper.fail('no new version')

lines[lineNumber] = line.replace(oldVersion, newVersion)


if lines[lineNumber] in open("./bot/previouslyFailedTheBuild", "r").read():
    helper.fail('previouslyFailedTheBuild')

open(gradleFile, "w").writelines(lines)

unitTestResult = helper.runUnitTests()
if 'BUILD SUCCESSFUL' not in unitTestResult:
    open("./bot/previouslyFailedTheBuild", 'a').write(lines[lineNumber] + '\n')
    helper.fail('unit test not success')

helper.command("""
git add {2}
git commit -m 'gradle-bot: {0} -> {1}'
""".format(helper.cleanUpLine(line), helper.cleanUpLine(lines[lineNumber]), gradleFile))

print('SUCCESS_NEW_VERSION_COMMITED')

# helper.commandPrint("""
# # git status
# git diff
# """)

File: code/processLine.js
const helper = require('./helper')
const cache = require('./cache')
const fs = require('fs/promises')

async function processLine({ lines, line, lineNumber, parts, gradleFile, repoDirectory, botDirectory }) {
    const oldVersion = helper.matchVersion(line, true)
    if (!oldVersion) {
        console.error('no old version')
        return
    }
    const tempVersion = helper.getTempVersion(oldVersion, parts)
    if (!tempVersion) {
        console.error('no temp version')
        return
    }
    console.log('tempVersion', tempVersion)
    
    let newVersion = null

    if (/testImplementation|implementation|compile/.test(line)) {
        const newDependencies = await helper.readLines(`${botDirectory}/parts${parts}.txt`)
        const matchingLine = helper.lookForFirstOccuranceOfPackage(newDependencies, line)
        newVersion = helper.matchVersion(matchingLine, false)
    } else {
        const getDependencies = async () => (await cache
            .gradleDependencies(repoDirectory, gradleFile, botDirectory, ''))
            .split('\n')
        const oldDependencies = await getDependencies()
        lines[lineNumber] = line.replace(oldVersion, tempVersion)
        await helper.writeLines(gradleFile, lines)
        const newDependencies = await getDependencies()
        const diff = helper.lookForFirstDiff(oldDependencies, newDependencies, 3)
        if (diff != null) {
            newVersion = helper.matchVersion(diff[1], false)
        }
    }
    console.log('newVersion', newVersion)
    if (!helper.isNewVersionActuallyNewer(newVersion, oldVersion)) {
        console.error('no new version')
        return
    }
    
    lines[lineNumber] = line.replace(oldVersion, newVersion)
    helper.writeLines(gradleFile, lines)
    
    const branchName = helper.branchName(lines[lineNumber])
    const remoteBranchAlreadyExists = await helper.remoteBranchAlreadyExists(repoDirectory, branchName)
    const gitCheckoutSuccess = await helper.gitCheckout(repoDirectory, branchName)
    
    console.log('branchName', branchName)
    if (remoteBranchAlreadyExists || !gitCheckoutSuccess) {
        console.error('branch already exists')
        return
    }

    await helper.execPrint(`
        cd ${repoDirectory}
        git add ${gradleFile}
        git commit -m 'gradle-bot: ${helper.cleanUpLine(line)} -> ${helper.cleanUpLine(lines[lineNumber])}'
        git log -1
        git push github
        git reset HEAD~1
    `)
    global.gitPushCount++
    console.log('SUCCESS_NEW_VERSION_COMMITED')

    if (global.gitPushCount >= 20) {
        // Just so we don't spam the CI build
        throw 'gitPushCount' + global.gitPushCount
    }
}
    
module.exports = processLine
